### PR TITLE
Using [NotNull] and 'nameof' operator

### PR DESCRIPTION
**Pull request by hishamco**

### CODE REVIEW

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs [hunk 1]
```diff
@@ -526,13 +527,8 @@ protected override AuthenticationTicket AuthenticateCore()
         /// <param name="nonce">the nonce to remember.</param>
         /// <remarks><see cref="HttpResponse.Cookies.Append"/>is called to add a cookie with the name: 'OpenIdConnectAuthenticationDefaults.Nonce + <see cref="OpenIdConnectAuthenticationOptions.StringDataFormat.Protect"/>(nonce)'.
         /// The value of the cookie is: "N".</remarks>
-        private void WriteNonceCookie(string nonce)
+        private void WriteNonceCookie([NotNull] string nonce)
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: NotNull doesn't quite replace IsNullOrEmpty. (Though we've been told to avoid IsNullOrWhitespace).

File: src/Microsoft.AspNet.Authentication/CertificateSubjectKeyIdentifierValidator.cs [hunk 2]
```diff
@@ -27,7 +27,7 @@ public CertificateSubjectKeyIdentifierValidator([NotNull] IEnumerable<string> va
 
             if (_validSubjectKeyIdentifiers.Count == 0)
             {
-                throw new ArgumentOutOfRangeException("validSubjectKeyIdentifiers");
+                throw new ArgumentOutOfRangeException(nameof(_validSubjectKeyIdentifiers));
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication/CertificateSubjectKeyIdentifierValidator.cs]: Shouldn't use a private variable name.

Comment (hishamco) [inline src/Microsoft.AspNet.Authentication/CertificateSubjectKeyIdentifierValidator.cs]: @Tratcher shall i use ```_validSubjectKeyIdentifiers.Count``` instead

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs (hunk 1 above)
Comment (hishamco) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: what should we use in this case

File: src/Microsoft.AspNet.Authentication/CertificateSubjectKeyIdentifierValidator.cs (hunk 2 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication/CertificateSubjectKeyIdentifierValidator.cs]: Use the original method param `validSubjectKeyIdentifiers`

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs (hunk 1 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: This should just check for IsNullOrEmpty.

File: src/Microsoft.AspNet.Authentication/CertificateSubjectKeyIdentifierValidator.cs (hunk 2 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Authentication/CertificateSubjectKeyIdentifierValidator.cs]: @Tratcher should this be a more general `ArgumentException` since the provided `string` isn't out of anybody's range?

Comment (Tratcher) [inline src/Microsoft.AspNet.Authentication/CertificateSubjectKeyIdentifierValidator.cs]: Hmm, yes.

Comment (hishamco) [inline src/Microsoft.AspNet.Authentication/CertificateSubjectKeyIdentifierValidator.cs]: @Tratcher I update the PR

File: src/Microsoft.AspNet.Authentication.OAuthBearer/OAuthBearerAuthenticationOptions.cs [hunk 3]
```diff
@@ -140,15 +140,10 @@ public TokenValidationParameters TokenValidationParameters
             {
                 return _tokenValidationParameters;
             }
-
+            [param: NotNull]
```
Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication.OAuthBearer/OAuthBearerAuthenticationOptions.cs]: FYI, this property will be simplified in another PR: https://github.com/aspnet/Security/pull/276/files#diff-377cf3927416f05a94fa6f3a47471549R114

File: src/Microsoft.AspNet.Authentication.OAuthBearer/OAuthBearerAuthenticationOptions.cs [hunk 4]
```diff
@@ -140,15 +140,10 @@ public TokenValidationParameters TokenValidationParameters
...
             set
             {
-                if (value == null)
-                {
-                    throw new ArgumentNullException("TokenValidationParameters");
-                }
-
-                _tokenValidationParameters = value;
+                 _tokenValidationParameters = value;
```
Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication.OAuthBearer/OAuthBearerAuthenticationOptions.cs]: Why do you need an extra space?

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationOptions.cs [hunk 5]
```diff
@@ -252,13 +252,9 @@ public ISecureDataFormat<AuthenticationProperties> StateDataFormat
             {
                 return _stateDataFormat;
             }
+            [param: NotNull]
```
Comment (PinpointTownes) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationOptions.cs]: Same remark as above, these properties will all be converted to automatic properties :smile:

File: src/Microsoft.AspNet.Authentication.OAuthBearer/OAuthBearerAuthenticationOptions.cs (hunk 4 above)
Comment (hishamco) [inline src/Microsoft.AspNet.Authentication.OAuthBearer/OAuthBearerAuthenticationOptions.cs]: Fixed, thanks @PinpointTownes

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationOptions.cs (hunk 5 above)
Comment (hishamco) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationOptions.cs]: @PinpointTownes what I have seen here https://github.com/aspnet/Common/blob/dev/src/Microsoft.Framework.NotNullAttribute.Sources/NotNullAttribute.cs#L8 is ```[NotNull]``` usage is ```Parameter``` not ```Property```

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: @hishamco looks like you haven't reverted changes to this file.

Would be nice to switch restored line 531 from `IsNullOrWhiteSpace()` to `IsNullOrEmpty()` and use `nameof` on line 533 at this time.

At some point should also switch from `ArgumentNullException` to `ArgumentException` since an empty `string` isn't `null`. A few projects in a few repos have an `ArgumentCannotBeNullOrEmpty` resource for this case. Should add one (with matching wording) to this project. Again, not necessarily part of this PR.

Comment (hishamco) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: The PR has been updated, thanks @dougbu

**Comment by dougbu** [comment]
LGTM

File: src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs [hunk 6]
```diff
@@ -528,11 +528,11 @@ protected override AuthenticationTicket AuthenticateCore()
...
         private void WriteNonceCookie(string nonce)
         {
-            if (string.IsNullOrWhiteSpace(nonce))
+            if (string.IsNullOrEmpty(nonce))
             {
-                throw new ArgumentNullException("nonce");
+                throw new ArgumentNullException(nameof(nonce));
             }
-
+
```
Comment (Eilon) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: Whitespace :frowning:

Comment (hishamco) [inline src/Microsoft.AspNet.Authentication.OpenIdConnect/OpenIdConnectAuthenticationHandler.cs]: Added by mistake in the latest commit, it got fixed

**Comment by Eilon** [comment]
Can you squash the commits in your fork?

**Comment by hishamco** [comment]
@Eilon I squash all the commits into two main commits

**Comment by Eilon** [comment]
@Tratcher please merge when you have a chance.

**Comment by Tratcher** [comment]
Merged.
